File: NowPlaying/ObjectiveC/Classes/ApplePosterDownloader.m
// Copyright (C) 2008 Cyrus Najmabadi
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 2 of the License, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program; if not, write to the Free Software Foundation, Inc., 51
// Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

#import "ApplePosterDownloader.h"

#import "Application.h"
#import "DifferenceEngine.h"
#import "Movie.h"
#import "Utilities.h"

@implementation ApplePosterDownloader

static NSDictionary* movieNameToPosterMap = nil;

+ (void) createMap {
    if (movieNameToPosterMap != nil) {
        return;
    }

    NSString* url = [NSString stringWithFormat:@"http://%@.appspot.com/LookupPosterListings", [Application host]];
    NSString* index = [Utilities stringWithContentsOfAddress:url important:NO];
    if (index == nil) {
        return;
    }

    NSMutableDictionary* result = [NSMutableDictionary dictionary];

    NSArray* rows = [index componentsSeparatedByString:@"\n"];
    for (NSString* row in rows) {
        NSArray* columns = [row componentsSeparatedByString:@"\t"];

        if (columns.count >= 2) {
            NSString* movieName = [Movie makeCanonical:[columns objectAtIndex:0]];
            NSString* posterUrl = [columns objectAtIndex:1];

            [result setObject:posterUrl forKey:movieName];
        }
    }

    movieNameToPosterMap = result;
    [movieNameToPosterMap retain];
}

+ (NSData*) download:(Movie*) movie {
    [self createMap];
    if (movieNameToPosterMap == nil) {
        return nil;
    }

    NSString* key = [[DifferenceEngine engine] findClosestMatch:movie.canonicalTitle inArray:movieNameToPosterMap.allKeys];
    if (key == nil) {
        return nil;
    }

    NSString* posterUrl = [movieNameToPosterMap objectForKey:key];
    return [Utilities dataWithContentsOfAddress:posterUrl important:NO];
}

@end

File: Android/1.3/src/org/metasyntactic/caches/posters/ApplePosterDownloader.java
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package org.metasyntactic.caches.posters;

import org.metasyntactic.NowPlayingApplication;
import org.metasyntactic.data.Movie;
import org.metasyntactic.utilities.NetworkUtilities;
import static org.metasyntactic.utilities.StringUtilities.isNullOrEmpty;
import org.metasyntactic.utilities.difference.EditDistance;

import java.util.HashMap;
import java.util.Map;

public class ApplePosterDownloader {
  private static Map<String, String> movieNameToPosterMap;

  private ApplePosterDownloader() {

  }

  public static byte[] download(final Movie movie) {
    createMap();

    if (movieNameToPosterMap == null) {
      return null;
    }

    final String key = EditDistance.findClosestMatch(movie.getCanonicalTitle(), movieNameToPosterMap.keySet());
    if (key == null) {
      return null;
    }

    final String address = movieNameToPosterMap.get(key);
    return NetworkUtilities.download(address, false);
  }

  private static void createMap() {
    if (movieNameToPosterMap != null) {
      return;
    }

    final String index = NetworkUtilities.downloadString("http://" + NowPlayingApplication.host + ".appspot.com/LookupPosterListings", false);
    if (isNullOrEmpty(index)) {
      return;
    }

    final Map<String, String> result = new HashMap<String, String>();

    for (final String row : index.split("\n")) {
      final String[] columns = row.split("\t");
      if (columns.length >= 2) {
        final String movieName = Movie.makeCanonical(columns[0]);
        final String posterUrl = columns[1];

        result.put(movieName, posterUrl);
      }
    }

    movieNameToPosterMap = result;
  }

  public static void onLowMemory() {
    movieNameToPosterMap = null;
  }
}
